easybuild/easyconfigs/j/jax/jax-0.2.18-foss-2020b.eb

```diff
@@ -0,0 +1,75 @@
+# This file is an EasyBuild reciPY as per https://github.com/easybuilders/easybuild
+# Author: Denis Kristak
+easyblock = 'PythonBundle'
+
+name = 'jax'
+version = '0.2.18'
+
+homepage = 'https://pypi.python.org/pypi/jax'
+description = """Composable transformations of Python+NumPy programs: differentiate, vectorize, JIT to GPU/TPU, and
+more"""
+
+toolchain = {'name': 'foss', 'version': '2020b'}
+
+# downloading TF tarball for better control of what Bazel will use during installation of Jax
+local_tf_tarball_version = '2.5.0'
+
+builddependencies = [
+    ('Bazel', '3.7.2'),
+]
+
+dependencies = [
+    ('Python', '3.8.6'),
+    ('SciPy-bundle', '2020.11'),
+    ('Cython', '0.29.22'),
+    ('opt-einsum', '3.3.0'),
+    ('flatbuffers-python', '1.12'),
+]
+
+local_tf_builddir = "%%(builddir)s/tensorflow-%s" % local_tf_tarball_version
+
+local_jax_preinstallopts = "sed -i -e 's$pathToSed$%s$g' WORKSPACE && " % local_tf_builddir
+local_jax_preinstallopts += 'python build/build.py '
+local_jax_preinstallopts += '--bazel_startup_options="--output_user_root=%(builddir)s" '
+local_jax_preinstallopts += '--bazel_path="$EBROOTBAZEL/bin/bazel" '
+local_jax_preinstallopts += '--bazel_options=--subcommands '
+local_jax_preinstallopts += '--bazel_options=--jobs=1 --bazel_options=--action_env=PYTHONPATH '
+local_jax_preinstallopts += '--bazel_options=--action_env=EBPYTHONPREFIXES && '
+local_jax_preinstallopts += 'pip install --prefix %(installdir)s dist/*.whl && '
+
+exts_list = [
+    ('opt-einsum', '3.3.0', {
+        'source_tmpl': 'opt_einsum-%(version)s.tar.gz',
+        'modulename': 'opt_einsum',
+    }),
+    ('absl-py', '0.13.0', {
+        'modulename': 'absl'
+    })]
+
+components = [
+    (name, version, {
+        'easyblock': 'PythonPackage',
```

Comment:
I'd suggest to make an extra component for jaxlib:
- Also use PythonPackage (basically C&P what you have here but adjust name and version [jaxlib has a different versioning])
- put the stuff from preinstallopts into buildcmd
- use `install_src = 'dist/*.whl'`

See https://github.com/easybuilders/easybuild-easyblocks/pull/2534 which then needs to be used.
The pip install command is actually a bit more involved and it makes sense to install jax and jaxlib separately as they are 2 packages which both would need sanity checks etc.

The jax part can then be trivial: Download from PyPI (default -> no source_urls) with IIRC default SOURCE_TAR_GZ template, no patches, no preinstallopts

Reply:
@Flamefire Can you clarify why you prefer separate components for `jaxlib` and `jax`?

Apparently both come from the same repository, but have different versioning schemes...

So if we split them up, rather than just using the `jaxlib` subdirectory in the `jax` release, we may be installing a combination that nobody else uses? Or am I missing something?

Also, why haven't I come up with "use one big repository for different tools, release with different versioning schemes" in my funny talk? It certainly ticks off the "cause confusion" requirement...

Reply:
I'm for separate, known versions of `jaxlib` and `jax` - so that we are using released versions of both - rather than whatever state `jaxlib` was in when for the version of `jax` we are using.

I also think we should be naming `jax-0.2.18-foss-2020b-jaxlib-0.1.70.eb`, or something similar, that makes it clear what versions of each we are installing.

For the combination we install - that should be versions of both released at about the same time - unless there is a known broken combination.

Reply:
>I'm for separate, known versions of jaxlib and jax - so that we are using released versions of both - rather than whatever state jaxlib was in when for the version of jax we are using.

Exactly. Also simply for the sanity checks. jaxlib should succeed without jax installed

And no I wouldn't suffix that. Same as with e.g. TF which as an absl-py as an extension and mentions that in the module file.

Reply:
We should only have a versionsuffix when we diverge for whatever reason. The default should be latest `jaxlib` with latest `jax`. If we diverge from that, we can discriminate with a versionsuffix?